relayer/src/link/relay_path.rs

```diff
@@ -1438,62 +1664,88 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
     /// waiting for the delay period to pass.
     /// Returns `None` if there is no operational data scheduled.
-    pub(crate) fn fetch_scheduled_operational_data(&self) -> Option<OperationalData> {
-        let odata = self
-            .src_operational_data
-            .pop_front()
-            .or_else(|| self.dst_operational_data.pop_front());
+    pub(crate) fn fetch_scheduled_operational_data(
+        &self,
+    ) -> Result<Option<OperationalData>, LinkError> {
+        let (delay_time_left, delay_blocks_left, odata) =
+            if let Some(odata) = self.src_operational_data.pop_front() {
+                let src_chain_time = self.src_time_latest()?;
+                let src_block_delay = self.src_conn_block_delay()?;
+                let src_latest_height = self.src_latest_height()?;
+                (
+                    self.conn_time_delay_elapsed(&odata, src_chain_time).err(),
+                    self.conn_block_delay_elapsed(&odata, src_block_delay, src_latest_height)
+                        .err(),
+                    Some(odata),
+                )
+            } else if let Some(odata) = self.dst_operational_data.pop_front() {
+                let dst_chain_time = self.dst_time_latest()?;
+                let dst_block_delay = self.dst_conn_block_delay()?;
+                let dst_latest_height = self.dst_latest_height()?;
+                (
+                    self.conn_time_delay_elapsed(&odata, dst_chain_time).err(),
+                    self.conn_block_delay_elapsed(&odata, dst_block_delay, dst_latest_height)
+                        .err(),
+                    Some(odata),
+                )
+            } else {
+                (None, None, None)
+            };
 
         if let Some(odata) = odata {
-            // Check if the delay period did not completely elapse
-            let delay_left = self
-                .channel
-                .connection_delay
-                .checked_sub(odata.scheduled_time.elapsed());
-
-            match delay_left {
-                None => info!(
+            match (delay_time_left, delay_blocks_left) {
+                (None, None) => info!(
                     "ready to fetch a scheduled op. data with batch of size {} targeting {}",
                     odata.batch.len(),
                     odata.target,
                 ),
-                Some(delay_left) => {
+                (Some(delay_time), _) => {
                     info!(
                         "waiting ({:?} left) for a scheduled op. data with batch of size {} targeting {}",
-                        delay_left,
+                        delay_time,
                         odata.batch.len(),
                         odata.target,
                     );
 
                     // Wait until the delay period passes
-                    thread::sleep(delay_left);
+                    thread::sleep(delay_time);
                 }
+                (None, Some(_blocks)) => {}
```

Comment:
It isn't clear to me how to handle cases where we have a remaining block delay.

Reply:
What situations would lead to there being remaining block delay?

Reply:
The block delay is calculated as ->
```
block_delay = time_delay / max_expected_time_per_block
```
So the `block_delay` is basically calculating the (minimum) number of blocks a chain would create in the `time_delay` duration. 
And the default value for `max_expected_time_per_block` in ibc-go is 30s. The rationale for choosing this value is described in detail in this comment -> https://github.com/cosmos/ibc-go/pull/171#discussion_r633870345 
> I see the block delay as handling the edge case where the receiving chain slows down unexpectedly (...)

So, `max_expected_time_per_block` is not the same as average block-time of the chain and is usually larger. This means that the calculated `block_delay` is usually much smaller than the average number of blocks the chain would be able to create in that (delay) duration. So, IMHO it is quite unlikely that there would be a remaining `block_delay` (if everything is configured correctly) and it should be sufficient if the relayer simply handles the `time_delay`.

Reply:
> So, IMHO it is quite unlikely that there would be a remaining block_delay (if everything is configured correctly) and it should be sufficient if the relayer simply handles the time_delay.

We should probably wait `blocks * max_expected_time_per_block`. 

I think the reasoning behind this is to allow a minimum number of blocks to pass between the client update and the recv or timeout packets so client misbehavior is detected, evidence submitted, client is frozen. And as you pointed out in the ibc-go issue, it should be possible to handle also the case where chain slows down significantly. For example if we expect the  block time to be at most 30sec, with a connection delay of 2 min the block check requires 4 blocks between the update client and the packet. Even if the block after the update is > 2min. Without this check it would be possible to process the `recv_packet` before the evidence of client misbehavior (that freezes the client).

Reply:
Done https://github.com/informalsystems/ibc-rs/pull/1917/files#diff-2a5a700c0b1975ad9d64941997aa53e0b6e26e577fa8f1e9a2aba4abe7cf1f95R1625 👍